### ISSUE TITLE
[RFC] IPv6 support

### ISSUE THREAD

**Issue by nextgens** [Mailu/Mailu#2293]
The current ipv6 support in Mailu is broken (and has been for a very long time: see #2260, #2272 and #1789). This ticket aims to discuss what the final end-state should be.

Architecturally the choices are:
1. do ipv4 exclusively internally (get rid of SUBNET6)
  - pro: simple
  - con: won't make outbound ipv6 connections (partial ipv6 support)
2. dual-stack (what we are attempting to do but what has been broken for a very long time)
  - con: complicated, breaks silently and requires NAT
3. do ipv6 when requested internally (SUBNET or SUBNET6, not both)
  - con: requires NAT

When I say requires NAT, I mean we currently assign ULAs to SUBNET6 to prevent open-relays (in setup, not sure what helm does)... so any outbound connection will be NATed and may not have the right PTR record. Keep in mind that it's the smtp/postfix container that makes outbound connections "outside" from its own address (unless it's NATed).

**Comment by nextgens** [comment]
@Niduroki @peterpramb @ghostwheel42 @Diman0 let's discuss it here.

I've discussed it with @Diman0 on #mailu-dev and the current view is that we are fine with 1.9 having broken v6 support and if we improve it it will be on master, once a consensus has emerged.

**Comment by Niduroki** [comment]
I think a perfect solution would be able to work without NAT at all – NAT4, NAT64, or NAT66 – not all cloud runtimes support NAT66 from what I've gathered (notably podman, I guess? – I don't think Mailu supports other cloud runtimes?).

Meaning, that SUBNET could be a publicly routable IP-range (IPv4 *or* IPv6), and that internal access (authentication-less, that is) is only possible, if the sender-address is in `SUBNET`/`SUBNET6`.
(Since TCP handshakes make sure that the sender address isn't spoofed, this should be sound, too - at least when thinking about it for a minute)
Makes distributed, horizontally scaled, setups way more easy, too.

That's in a perfect world, at least - we'd definitely need to make sure, that every connection is TLS-encrypted in this scenario, too, and I'm not too certain, if that is a possibility in some of the backend connections, plus any misconfiguration would cause open-relays even in the IPv4 scenario.

SMTP rDNS/PTR being problematic could be fixed, by making sure, that the SMTP container gets assigned a fixed IP-address, when working without NAT. Same thing really with front, too, and its related `MX/A/AAAA` records

-----

Back to reality, though.

Scenario 2 isn't really anything, the longer I think about it. It's in between 1 and 3, but not quite any of them.

I'm 100% advocating for scenario 3, i.e. NAT66, rather than NAT64, which is what scenario 1 (and scenario 2, I guess?) does.

Docker IPv6 support is *really* uncertain, though. For 15 months now, the experimental `ip6tables` docker switch (required for doing NAT66) is in a bit of an uncertain space (https://github.com/moby/moby/issues/41754) and I don't have high hopes of this changing in the near future, leaving it as experimental, which isn't really great for peace of mind, to build an architectural decision upon.

Alternatively we could build NAT66 upon [robbertkl/docker-ipv6nat](https://github.com/robbertkl/docker-ipv6nat), which basically does the same thing as `dockerd --experimental --ip6tables --fixed-cidr-v6 ...`, and is listed in current mailu the documentation, *but* will eventually be replaced by named `dockerd` configuration, and is thus pretty dead (*and* is combined with more configuration, in comparison to the named `dockerd` config).

All of this somewhat stacks the cards against me and scenario 3, due to dockers unwillingness to do anything about IPv6.

@peterpramb: You're one of the other brave souls using IPv6 in production, and have also been against removing `SUBNET6`, but spelling out all of this, it seems like the only sane option. At least in the common docker, scenario …

**Comment by peterpramb** [comment]
Well, no offense intended, but the question is probably not which approach to choose best, but rather what is actually dictated by the container engines in use or supported (or their CNI, to be more exact).

I can only talk about `Podman`, as this is my use case, and not about `Docker`, `Kubernetes`, or whatever else. I'm aware that `Podman` is not mentioned anywhere as supported, I'm just trying to provide some input from another container engine's perspective to the discussion, nothing else.

I mentioned in #2260 (or tried to) that the internal (btw. that means bridged, so I'm talking about `rootfull` containers here; `rootless` containers are a different beast with their own network stack) Mailu network needs to be IPv6-enabled, not because I want it, but because it's required by `Podman`'s builtin CNI (which uses DNAT with Netfilter) to actually publish ports for IPv6:

| Internal Network | Public Interface | Published Port |
| :---: | :---: | :---: |
| IPv4 only | IPv4 only | IPv4 only |
| IPv4 only | IPv4+v6 | IPv4 only |
| IPv4+v6 | IPv4 only | not tested |
| IPv4+v6 | IPv4+v6 | IPv4+v6 |

In other words: No internal IPv6, no external IPv6 reachability (this applies to both ULA and public IPv6 addresses on the internal network, as I just tested). The same for outgoing container traffic as well.

So I can either have IPv4-only in and out, or dual stack for both, but nothing else between. That's why I asked not to remove any already existing `SUBNET6` support for now (as long as there is no better approach), as it would force me to drop IPv6 altogether.

**Comment by pr0ton11** [comment]
@peterpramb Do you have a write up how you got it working with podman?
I have no clue why, but my IPv6 attempts with docker so far did not work at all. At the moment I am failing to reach the front container with port publish (-p). It always runs into a TCP timeout and in the logs I don't see anything.
I started initially with just using IPv4 and adding IPv6 afterwards (that might be the issue?)
Thanks for any help

**Comment by peterpramb** [comment]
Unfortunately I never used Docker with IPv6, so no experience here.

In Podman AFAIR it simply worked, it does not handle IPv6 differently than IPv4. Just make sure that:
1. The host itself has IPv6 reachability (no _external_ firewall blocks incoming/outgoing IPv6 traffic).
2. No _local_ firewall blocks _forwarding_ traffic to the container/published ports.
3. The container network to use (default network, if none given) is IPv6 enabled (you have to specify an IPv6 range when creating a new network, e.g. `fd00:89:1::/64` for the first custom network `10.89.1.0/24`. The default network however needs to be manually reconfigured).
4. The IPv6 address for the published port (if any) is specified in the correct format (`--publish=[1:2:3::a]:80:8080/tcp`).
5. When using `nftables` as the backend, no other tables have blocking _forward_ rules for IPv6 in general or for the intended container/published ports.

**Comment by pr0ton11** [comment]
Thanks for the help, I got it working as well with Podman and Proxy-V2. I will create a writeup for other people to make it easier for them to use IPv6.

**Comment by lel-amri** [comment]
I'm running an IPv6-only Kubernetes cluster with a NAT64 gateway. I'm using Cilium with the kube-proxy replacement as the CNI.

My opinion on the architectural choices is:

1. Would completely prevent a meaningful use of IPv6 because of the difference in addresses space size between IPv6 and IPv4, unless one wants to manage a nasty NAT46 configuration for Mailu specifically. It would also force users that only want IPv6 to do IPv4.

2. Dual-stack is a nightmare.

3. I'm not sure that I follow this one properly. It doesn't say that IPv4 is disabled when IPv6 is explicitly requested. To me it falls in case 2: dual-stack is a nightmare.

I believe the only sane choice is to forget about dual-stack altogether. Do either one or the other, not both, and offload the burden of reconciling the IPv4 <-> IPv6 boundary to the administrator.

I tinkered for one year to get a network that I like, this network is IPv6 only with a NAT64 gateway. What I identifies _now_ as the issue I had _then_ is the dual-stack IP. I have the feeling supporting this is not worth the effort: administrators that know how to manage an IPv6-only network know how to make it talk with the IPv4 Internet. And they are part of the superset "administrators that know how to do dual-stack IP and haven't lost their mental sanity yet".

I'm not omniscient, so I may be ignoring some security features that are based on IP addresses (TLS CN ?), or mail-specific protocols, or DNS witchcraft. I'd be glad if anyone could bring them to my attention.

**Comment by nextgens** [comment]
So docker 27 now officially supports ipv6
[https://docs.docker.com/engine/release-notes/27/#ipv6](https://docs.docker.com/engine/release-notes/27/#ipv6)
